### PR TITLE
Fixing Remove in ForAll return type issue

### PR DESCRIPTION
**Pull request by anderson-joyle**
Issue https://github.com/microsoft/Power-Fx/issues/1685.

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1973]: >&& !tableType.FieldNames.Any() [](http://example.com/codeflow?start=69&length=30)

This can be merged into the previous if statement (`...arg is BlankValue bv && (fieldCount == 0 || fieldCount > 1))`)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1967]: >tableType.FieldNames.Count() [](http://example.com/codeflow?start=71&length=28)

Can be cached outside the loop (e.g., `var fieldCount = tableType.FieldNames.Count();` after the declaration of tableType) since the value doesn't change

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1972]: >Coelesce [](http://example.com/codeflow?start=69&length=8)

[typo] `Coalesce`

Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1972]: >Handle the Unknown return type functions [](http://example.com/codeflow?start=19&length=40)

It's not really an unknown return type - we do know the type. It's mostly untyped blank values.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (CarlosFigueiraMSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1967]: >tableType.FieldNames.Count() > 1 [](http://example.com/codeflow?start=71&length=32)

Do we need this restriction at all? For example, today this throws an exception: `ForAll([1,2,3], If(Value <> 2, {a:Value}))`

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt:8]: >able(Blank()) [](http://example.com/codeflow?start=1&length=13)

We should add this test, to show that `list` was properly updated:

    >> list
    Table({ID:2,Name:"Two"})

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt:13]: >ForAll(x,Remove(x, ThisRecord)) [](http://example.com/codeflow?start=3&length=31)

This is dangerous - we should not let the collection being 'looped' to be updated. Similar to the exception thrown in C# when we do something like

    var coll = new List<int>() { 1, 2, 3 };
    foreach (var i in coll)
    {
        coll.Add(i);
    }

or

    var coll = new List<int>() { 1, 2, 3 };
    foreach (var i in coll)
    {
        coll.Remove(i);
    }

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1967]: I would need to investigate this situation as part of another issue.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt
Comment (anderson-joyle) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt:13]: Issue [https://github.com/microsoft/Power-Fx/issues/1696](https://github.com/microsoft/Power-Fx/issues/1696) will address this.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (CarlosFigueiraMSFT) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs]: This is actually the same issue - blank values that are not handled properly.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt
Comment (CarlosFigueiraMSFT) [reply src/tests/Microsoft.PowerFx.Interpreter.Tests/MutationScripts/Remove.txt]: In any case, let's not add this test. Tests should always show what is expected in the product - and that is not.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt:120]: This is wrong - the expected result should be `Table({a:1},Blank(),{a:3})`. The type of the second argument in ForAll is a record, so its result should not be wrapped.

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1961]: This file looks like it has a ton of whitespace changes - was that intentional? (it makes it harder to see the changes you actually made; and it will interfere with git blame)

File: src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1969]: `fieldCounter >= 0` ?

Or just:
`tableType.FieldNames.Any()`  ?

Comment (anderson-joyle) [reply src/libraries/Microsoft.PowerFx.Interpreter/Functions/Library.cs:1969]: The original logic was checking if the field counter was different from 1. Originally I decided not to change much but turns out this the root of the problem.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt:126]: This is covering the same codepath as the test on line 133; your call whether to keep it or remove this one

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt:123]: This is covering the same codepath as the test on line 129; your call whether to keep it or remove this one

File: src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt
Comment (CarlosFigueiraMSFT) [inline src/tests/Microsoft.PowerFx.Core.Tests/ExpressionTestCases/ForAll.txt:125]: >{CountRows(Value)} [](http://example.com/codeflow?start=160&length=18)

[minor] space between the number

**Review by CarlosFigueiraMSFT — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**
